File: taskauthor/ModifiedDijkstra.c
#include <stdio.h>
#include <string.h>

#define INF 1000000000

int i, j, V, n, w, counter, Q, s, t, d, u, vv, w_u_vv;
int AdjList[1000][1000], weight[1000][1000];
int sz[1000], dist[1000];

struct p {
  int d, u;
};
struct p front;

//====================================================
// Start of Binary Heap implementation  
struct p A[1000];
int BinaryHeapSize;

int smaller(struct p *a, struct p *b) {
  if (a->d != b->d)
    return a->d < b->d;
  else
    return a->u < b->u;
}

int parent(int i) { return i >> 1; } // shortcut for i / 2, round down

int left(int i) { return i << 1; } // shortcut for 2 * i

int right(int i) { return (i << 1) + 1; } // shortcut for 2 * i + 1

void Initialize() {
  A[0].d = A[0].u = 0; // dummy
  BinaryHeapSize = 0;
}

void shiftUp(int i) {
  struct p temp;
  while (i > 1 && smaller(&A[i], &A[parent(i)])) {
    temp = A[i];
    A[i] = A[parent(i)];
    A[parent(i)] = temp;
    i = parent(i);
  }
}

void Enqueue(int _d, int _u) {
  BinaryHeapSize++;
  A[BinaryHeapSize].d = _d;
  A[BinaryHeapSize].u = _u;
  shiftUp(BinaryHeapSize);
}

void shiftDown(int i) {
  struct p minV, temp;
  int min_id;
  while (i <= BinaryHeapSize) {
    minV = A[i]; min_id = i;
    if (left(i) <= BinaryHeapSize && smaller(&A[left(i)], &minV)) {
      minV = A[left(i)];
      min_id = left(i);
    }
    if (right(i) <= BinaryHeapSize && smaller(&A[right(i)], &minV)) {
      minV = A[right(i)];
      min_id = right(i);
    }

    if (min_id != i) {
      temp = A[i];
      A[i] = A[min_id];
      A[min_id] = temp;
      i = min_id;
    }
    else
      break;
  }
}
  
struct p Dequeue() {
  struct p minV = A[1];
  A[1] = A[BinaryHeapSize];
  BinaryHeapSize--; // virtual decrease
  shiftDown(1);
  return minV;
}

int isEmpty() { return BinaryHeapSize == 0; }
// End of Binary Heap implementation  
//====================================================

int main() {
  scanf("%d", &V);
  for (i = 0; i < V; i++) {
    scanf("%d", &n);
    while (n--) {
      scanf("%d %d", &j, &w);
      AdjList[i][sz[i]] = j;
      weight[i][sz[i]] = w;
      sz[i]++;
    }
  }

  counter = 0;
  scanf("%d", &Q);
  while (Q--) {
    scanf("%d %d", &s, &t);

    for (i = 0; i < V; i++) dist[i] = INF;
    dist[s] = 0;
    Initialize();
    Enqueue(0, s);
    while (!isEmpty()) {
      counter++;
      if (counter > 1000000) {
        printf("TLE because iteration counter > 1000000\n");
        return 1;
      }

      front = Dequeue();
      d = front.d; u = front.u;
      if (d == dist[u]) {
        for (j = 0; j < sz[u]; j++) {
          vv = AdjList[u][j]; w_u_vv = weight[u][j];
          if (dist[u] + w_u_vv < dist[vv]) {
            dist[vv] = dist[u] + w_u_vv;
            Enqueue(dist[vv], vv);
          }
        }
      }
    }

    printf("%d\n", dist[t]);
  }

  printf("The value of counter is: %d\n", counter);
  return 0;
}

File: taskauthor/ModifiedDijkstra.pas
type
  p = Record
    d, u : Longint;
  end;

var
  i, j, V, n, w, counter, Q, s, t, d, u, vv, w_u_vv : Longint;
  AdjList, weight : array[0..999,0..999] of Longint;
  sz, dist : array[0..999] of Longint;
  front : p;

//====================================================
// Start of Binary Heap implementation  
  A : array[0..999] of p;
  BinaryHeapSize : Longint;

function smaller(a : p; b : p) : Boolean;
begin
  if a.d <> b.d then begin
    if a.d < b.d then smaller := true else smaller := false;
  end
  else begin
    if a.u < b.u then smaller := true else smaller := false;
  end;
end;

function parent(i : Longint) : Longint;
begin
  parent := trunc(i / 2);
end;

function left(i : Longint) : Longint;
begin
  left := 2 * i;
end;

function right(i : Longint) : Longint;
begin
  right := 2 * i + 1;
end;
  
procedure Initialize;
begin
  A[0].d := 0;
  A[0].u := 0; // dummy
  BinaryHeapSize := 0;
end;

procedure shiftUp(i : Longint);
var
  temp : p;
begin
  while (i > 1) and (smaller(A[i], A[parent(i)])) do begin
    temp := A[i];
    A[i] := A[parent(i)];
    A[parent(i)] := temp;
    i := parent(i);
  end;
end;

procedure Enqueue(_d : Longint; _u : Longint);
begin
  inc(BinaryHeapSize);
  A[BinaryHeapSize].d := _d;
  A[BinaryHeapSize].u := _u;
  shiftUp(BinaryHeapSize);
end;

procedure shiftDown(i : Longint);
var
  minV, temp : p;
  min_id : Longint;
begin
  while i <= BinaryHeapSize do begin
    minV := A[i]; min_id := i;
    if (left(i) <= BinaryHeapSize) and (smaller(A[left(i)], minV)) then begin
      minV := A[left(i)];
      min_id := left(i);
    end;
    if (right(i) <= BinaryHeapSize) and (smaller(A[right(i)], minV)) then begin
      minV := A[right(i)];
      min_id := right(i);
    end;

    if min_id <> i then begin
      temp := A[i];
      A[i] := A[min_id];
      A[min_id] := temp;
      i := min_id;
    end
    else
      break;
  end;
end;

function Dequeue() : p;
var
  minV : p;
begin
  minV := A[1];
  A[1] := A[BinaryHeapSize];
  dec(BinaryHeapSize); // virtual decrease
  shiftDown(1);
  Dequeue := minV;
end;

function isEmpty() : Longint;
begin
  if BinaryHeapSize = 0 then isEmpty := 1 else isEmpty := 0;
end;
// End of Binary Heap implementation  
//====================================================

begin
  read(V);
  for i := 0 to V-1 do begin
    read(n);
    while n > 0 do begin
      dec(n);
      read(j, w);
      AdjList[i,sz[i]] := j;
      weight[i,sz[i]] := w;
      inc(sz[i]);
    end;
  end;
 
  counter := 0;
  read(Q);
  while Q > 0 do begin
    dec(Q);
    read(s, t);

    for i := 0 to V-1 do begin
      dist[i] := 1000000000;
    end;
    dist[s] := 0;
    Initialize();
    Enqueue(0, s); 
    while isEmpty() = 0 do begin
      inc(counter);
      if counter > 1000000 then begin
        writeln('TLE because iteration counter > 1000000');
        exit;
      end;

      front := Dequeue();
      d := front.d; u := front.u;
      if d = dist[u] then begin
        for j := 0 to sz[u]-1 do begin
          vv := AdjList[u,j];
          w_u_vv := weight[u,j];
          if dist[u] + w_u_vv < dist[vv] then begin
            dist[vv] := dist[u] + w_u_vv;
            Enqueue(dist[vv], vv);
          end;
        end;
      end;
    end;

    writeln(dist[t]);
  end;
 
  writeln('The value of counter is ', counter);
end.
